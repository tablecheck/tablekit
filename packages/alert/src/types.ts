import { BaseButtonProps } from '@tablecheck/tablekit-button';
import { HTMLAttributes } from 'react';

type Action = Omit<BaseButtonProps, 'children'> & {
  onClick: VoidFunction;
  text: string;
};

export enum Appearance {
  Primary = 'primary',
  Success = 'success',
  Warning = 'warning',
  Danger = 'danger'
}

export enum Position {
  Top = 'top',
  TopLeft = 'top-left',
  TopRight = 'top-right',
  Bottom = 'bottom',
  BottomLeft = 'bottom-left',
  BottomRight = 'bottom-right'
}

export type AlertProps = HTMLAttributes<HTMLDivElement> & {
  /** The array includes some object for create actions button in a footer.
   * Here you can use Button props. */
  actions?: Action[];

  /** The base styling to apply to the Alert. */
  appearance?: Appearance;

  /** The message to be shown to user */
  children?: JSX.Element;

  /** Add a classname to the wrapper. */
  className?: string;

  /** If 'false' (by default), the cross button will be hidden */
  isDismissible?: boolean;

  /** Should show Alert */
  isShow?: boolean;

  /** Close the message */
  onClose?: VoidFunction;

  /** Responsible for the position of the Alert on the screen. By default it's 'top' */
  position?: Position;

  /** Should close alert on click outside or not. By default it's 'false' */
  shouldCloseOnClickOutside?: boolean;

  /** Wheather the icon should be shown or not */
  shouldHideIcon?: boolean;

  /** Should use portal for alerts or not. By default it's 'false' */
  shouldUsePortal?: boolean;
};

export type AlertPortalProps = {
  children: JSX.Element;
  position?: Position;
  shouldUsePortal: boolean;
};

export type AlertFooterProps = {
  actions: Action[];
};
