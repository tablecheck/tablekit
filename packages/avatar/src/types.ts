import { Size } from '@tablecheck/tablekit-theme';
import { HTMLAttributes } from 'react';

export interface AvatarProps extends HTMLAttributes<HTMLDivElement> {
  /** the URL for the image */
  imageUrl?: string;

  /** the initials from the full name  to use */
  initials?: string;

  /** the full name to use */
  name?: string;

  size?: Size;
}
