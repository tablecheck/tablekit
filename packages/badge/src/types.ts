import { Size } from '@tablecheck/tablekit-theme';
import { HTMLAttributes, ReactNode } from 'react';

export enum BadgeType {
  Solid = 'solid',
  Outline = 'outline'
}

export type BadgeSize = Extract<Size, Size.Regular | Size.Small>;

export type BadgeProps = HTMLAttributes<HTMLDivElement> & {
  /** custom className to put on the Container element */
  className?: string;

  /** render custom children instead of value */
  children?: ReactNode;

  /** The maximum value to display. If value is 100, and max is 50, "50+" will be displayed */
  maxValue?: number;

  /** Defines the size of the badge, by default it's 'regular' */
  size?: BadgeSize;

  /** Defines the visual style the badge, by default it's 'outline' */
  type?: BadgeType;

  /** Use it if you need a flexible component width. */
  isFlexible?: boolean;

  /** The value displayed within the badge. */
  value?: number;
};

export type BadgeContainerProps = Omit<BadgeProps, 'value' | 'maxValue'>;
