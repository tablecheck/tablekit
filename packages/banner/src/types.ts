import { IconDefinition } from '@tablecheck/tablekit-icon';
import { Size } from '@tablecheck/tablekit-theme';
import { HTMLAttributes } from 'react';

export enum Appearance {
  Primary = 'primary',
  Success = 'success',
  Warning = 'warning',
  Danger = 'danger'
}

export type BannerSizes =
  | Size.Small
  | Size.Regular
  | Size.Regular2
  | Size.Large
  | Size.XLarge
  | Size.XXLarge;

interface Dismissible {
  /** Defines if banner can be dismissed by user. By default is `false` */
  isDismissible: true;
  /** Close button click handler (should be provided if banner is dismissible) */
  onClose: () => void;
}

interface NonDismissible {
  isDismissible?: false;
  onClose?: () => void;
}

export type BannerProps = (Dismissible | NonDismissible) &
  HTMLAttributes<HTMLDivElement> & {
    /** One of predefined banner Appearance */
    appearance?: Appearance;

    /** Add a classname to the banner. By default is `primary` */
    className?: string;

    /** One of @fortawesome icons */
    customIcon?: IconDefinition;

    /** Message to display */
    message: string;

    /** Size of a banner. By default is `regular` */
    size?: BannerSizes;
  };
