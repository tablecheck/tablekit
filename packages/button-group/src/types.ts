import { IconPropType } from '@tablecheck/tablekit-button';
import { ButtonHTMLAttributes, HTMLAttributes } from 'react';

export type ItemType = Partial<
  ButtonHTMLAttributes<HTMLAnchorElement | HTMLButtonElement> & {
    iconAfter: IconPropType;
    iconBefore: IconPropType;
  }
> & {
  label: string;
  value: string;
};

export type ButtonGroupPropsType = Omit<
  HTMLAttributes<HTMLDivElement>,
  'onChange'
> & {
  isInvalid?: boolean;
  isMulti?: boolean;
  isValueRequired?: boolean;
  items: ItemType[];
  onChange: (newValues: string[]) => void;
  values?: string[];
  width?: string;
};
