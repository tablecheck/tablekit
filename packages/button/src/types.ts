import { MediaQuery } from '@tablecheck/tablekit-utils';
import * as React from 'react';

export enum Appearance {
  Disabled = 'disabled',
  Subtle = 'subtle',
  Solid = 'solid',
  Outline = 'outline',
  Inverse = 'inverse',
  SubtleOutline = 'subtle-outline',
  Vertical = 'vertical',
  DisabledVertical = 'disabled-vertical'
}

export enum Color {
  Primary = 'primary',
  Ghost = 'ghost',
  Success = 'success',
  Warning = 'warning',
  Danger = 'danger'
}

export type IconPropType = JSX.Element | ((props: unknown) => React.ReactNode);

export enum ButtonSize {
  Small = 'small',
  Regular = 'regular',
  Regular2 = 'regular2',
  Large = 'large'
}

export enum ButtonShape {
  Sharp = 'sharp',
  Rounded = 'rounded',
  Circular = 'circular'
}

export interface BaseButtonProps {
  children?: React.ReactNode;

  /** The base styling to apply to the button. */
  appearance?: Appearance;
  /** apply different "theme" colors to the button appearance */
  color?: Color;

  /** The base styling to apply to the button border radius. */
  shape?: ButtonShape;

  /** Places an icon within the button, after the button's text. */
  iconAfter?: IconPropType;

  /** Places an icon within the button, before the button's text. */
  iconBefore?: IconPropType;

  /** Set if the button is disabled. */
  isDisabled?: boolean;

  /** Change the style to indicate the button action is processing, also disables onClick. */
  isLoading?: boolean;

  isInteractive?: boolean;

  /** Option to fit button width to its parent width. Also sets large height if enabled. */
  shouldFitContainer?: boolean;

  /** Set the size of the button. */
  size?: MediaQuery<ButtonSize>;

  /** Set the type of the button. */
  type?: 'submit' | 'reset' | 'button';
}

type ConflictProps = 'color';

export type AnchorButtonProps = BaseButtonProps &
  Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, ConflictProps>;
export type ButtonProps = BaseButtonProps &
  Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, ConflictProps>;

export type ComponentButtonProps<Component extends React.ElementType> =
  BaseButtonProps & {
    /** Optional function to be used for rendering links. Receives `href` and possibly `target`
     * as props. */
    component: Component | undefined;
  } & Omit<
      React.ComponentPropsWithoutRef<Component>,
      'component' | ConflictProps
    >;

export type DerivedButtonProps = BaseButtonProps &
  React.DOMAttributes<HTMLAnchorElement | HTMLButtonElement> & {
    // derived from isDisabled
    disabled?: boolean;
    fit?: boolean;
    // passed from button implementation
    isOnlyChild?: boolean;
    color: Color;
  };
