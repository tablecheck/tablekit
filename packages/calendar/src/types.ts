export type { DateObj, Calendar } from 'dayzed';

export type SelectedRange = undefined | Date | [Date, Date] | [Date];
