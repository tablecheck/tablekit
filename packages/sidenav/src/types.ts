import { IconDefinition } from '@tablecheck/tablekit-icon';
import { ItemProps } from '@tablecheck/tablekit-item';
import { Size } from '@tablecheck/tablekit-theme';
import { ReactNode, ElementType } from 'react';

export enum SideNavState {
  Collapsed = 'collapsed',
  Expanded = 'expanded'
}

export enum NavItemAppearance {
  Default = 'default',
  Selected = 'selected',
  Disabled = 'disabled'
}

export interface SideNavContextState {
  isExpanded: boolean;
  size: Size.Regular | Size.Large;
  state: SideNavState;
  toggleSideNav: () => void;
  toggleStateLabels: {
    collapse: string;
    expand: string;
  };
  shouldHideToggle: boolean;
}

export interface SideNavProviderProps
  extends Partial<Omit<SideNavContextState, 'toggleSideNav'>> {
  children: ReactNode;
}

export interface BaseSideNavItemProps {
  /** one of @fortawesome icons */
  icon?: React.ReactNode | IconDefinition;

  /** Causes the item to appear in a disabled state and click behavior will not be triggered. */
  isDisabled?: boolean;

  /** Indicates that navigation item is a child */
  isNested?: boolean;

  /** Causes the item to appear with a persistent selected background state. */
  isSelected?: boolean;

  /** Defines the size of the item */
  size?: SideNavContextState['size'];
}

export type SideNavItemProps<Component extends ElementType> =
  ItemProps<Component> & BaseSideNavItemProps;

export type SideNavSectionProps<Component extends ElementType> =
  ItemProps<Component> & {
    /** number of section children
     * (if not passed there would be no transition on expand/collapse section)
     * */
    childrenCount?: number;

    /** one of @fortawesome icons */
    icon?: React.ReactNode | IconDefinition;

    /** Causes the item to appear in a disabled state and click behavior will not be triggered. */
    isDisabled?: boolean;

    /** Shows whether navigation section is expanded */
    isOpened?: boolean;

    /** Causes the item to appear with a persistent selected background state. */
    isSelected?: boolean;

    /** Label to show on sidebar */
    label?: string;
  };
