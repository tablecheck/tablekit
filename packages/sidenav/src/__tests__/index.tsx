import { mount } from '@tablecheck/tablekit-enzyme';
import '@tablecheck/tablekit-free-icon-config';
import { HTMLAttributes } from 'react';

import { SideNav, SideNavProvider } from '../index';
import { SideNavProviderProps } from '../types';

beforeEach(() => {
  global.console.error = jest.fn();
});

const SideBar = (
  props: Omit<SideNavProviderProps, 'children'> & HTMLAttributes<HTMLDivElement>
) => (
  <SideNavProvider {...props}>
    <SideNav {...props} />
  </SideNavProvider>
);

describe('Sidenav', () => {
  test('should be possible to create a component', () => {
    const wrapper = mount(<SideBar />);
    expect(wrapper).not.toBeUndefined();
  });

  test('should correctly apply styling', () => {
    expect(mount(<SideBar />)).toMatchSnapshot();
  });

  test('should allow passing of custom className', () => {
    expect(mount(<SideBar className="test-class" />)).toMatchSnapshot();
  });
});
