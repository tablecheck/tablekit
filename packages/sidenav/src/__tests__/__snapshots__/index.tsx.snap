// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidenav should allow passing of custom className 1`] = `
.emotion-0 {
  border-right: 1px solid rgba(0, 0, 0, 0.11);
  background-color: white;
  color: #323232;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-transition: all 0.3s cubic-bezier(0.77, 0.2, 0.05, 1);
  transition: all 0.3s cubic-bezier(0.77, 0.2, 0.05, 1);
  z-index: 300;
  max-width: 201px;
  min-width: 201px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  min-height: 0;
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-top: 1px solid rgba(0, 0, 0, 0.11);
  padding: 8px 20px;
  min-height: 40px;
  max-height: 40px;
  width: 100%;
}

.emotion-2:hover {
  background-color: rgba(0, 0, 0, 0.05);
}

.emotion-4 {
  cursor: pointer;
  overflow: hidden;
  padding-left: 8px;
  text-overflow: ellipsis;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
}

<SideBar
  className="test-class"
>
  <SideNavProvider
    className="test-class"
  >
    <SideNav
      className="test-class"
    >
      <SideNavBody
        className="test-class"
        isExpanded={true}
        size="regular"
      >
        <div
          className="test-class emotion-0 emotion-1"
          size="regular"
        >
          <SideNavFooter
            isExpanded={true}
            size="regular"
            toggleSideNav={[Function]}
            toggleStateLabels={
              Object {
                "collapse": "Collapse",
                "expand": "Expand",
              }
            }
          >
            <FooterWrapper
              isExpanded={true}
              size="regular"
              toggleSideNav={[Function]}
              tooltipContent="Expand"
            >
              <SideNavFooterContainer
                onClick={[Function]}
                size="regular"
              >
                <div
                  className="emotion-2 emotion-3"
                  onClick={[Function]}
                  size="regular"
                >
                  <Icon
                    icon="fas-chevron-left"
                    size="regular"
                  >
                    icon-chevron-left
                  </Icon>
                  <SideNavFooterLabel
                    size="regular"
                  >
                    <div
                      className="emotion-4 emotion-5"
                      size="regular"
                    >
                      Collapse
                    </div>
                  </SideNavFooterLabel>
                </div>
              </SideNavFooterContainer>
            </FooterWrapper>
          </SideNavFooter>
        </div>
      </SideNavBody>
    </SideNav>
  </SideNavProvider>
</SideBar>
`;

exports[`Sidenav should correctly apply styling 1`] = `
.emotion-0 {
  border-right: 1px solid rgba(0, 0, 0, 0.11);
  background-color: white;
  color: #323232;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-transition: all 0.3s cubic-bezier(0.77, 0.2, 0.05, 1);
  transition: all 0.3s cubic-bezier(0.77, 0.2, 0.05, 1);
  z-index: 300;
  max-width: 201px;
  min-width: 201px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  min-height: 0;
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-top: 1px solid rgba(0, 0, 0, 0.11);
  padding: 8px 20px;
  min-height: 40px;
  max-height: 40px;
  width: 100%;
}

.emotion-2:hover {
  background-color: rgba(0, 0, 0, 0.05);
}

.emotion-4 {
  cursor: pointer;
  overflow: hidden;
  padding-left: 8px;
  text-overflow: ellipsis;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
}

<SideBar>
  <SideNavProvider>
    <SideNav>
      <SideNavBody
        isExpanded={true}
        size="regular"
      >
        <div
          className="emotion-0 emotion-1"
          size="regular"
        >
          <SideNavFooter
            isExpanded={true}
            size="regular"
            toggleSideNav={[Function]}
            toggleStateLabels={
              Object {
                "collapse": "Collapse",
                "expand": "Expand",
              }
            }
          >
            <FooterWrapper
              isExpanded={true}
              size="regular"
              toggleSideNav={[Function]}
              tooltipContent="Expand"
            >
              <SideNavFooterContainer
                onClick={[Function]}
                size="regular"
              >
                <div
                  className="emotion-2 emotion-3"
                  onClick={[Function]}
                  size="regular"
                >
                  <Icon
                    icon="fas-chevron-left"
                    size="regular"
                  >
                    icon-chevron-left
                  </Icon>
                  <SideNavFooterLabel
                    size="regular"
                  >
                    <div
                      className="emotion-4 emotion-5"
                      size="regular"
                    >
                      Collapse
                    </div>
                  </SideNavFooterLabel>
                </div>
              </SideNavFooterContainer>
            </FooterWrapper>
          </SideNavFooter>
        </div>
      </SideNavBody>
    </SideNav>
  </SideNavProvider>
</SideBar>
`;
